### PR TITLE
Add new Engine API to accept multiple SyncActivity request

### PR DESCRIPTION
**Pull request by xwduan**
**What changed?**
Add new Engine API to accept multiple SyncActivity request

**Why?**
As a building block for batching SyncActivityTask in replication stack

**How did you test it?**
Unit test.

**Potential risks**
Replication stack will fail

**Is hotfix candidate?**
no

### CODE REVIEW

File: service/history/ndc/activity_state_replicator_test.go [hunk 1]
```diff
@@ -608,6 +608,43 @@ func (s *activityReplicatorStateSuite) TestSyncActivity_WorkflowNotFound() {
...
+				Clusters: []string{
+					cluster.TestCurrentClusterName,
+					cluster.TestAlternativeClusterName,
+				},
+			},
+			version,
+		), nil,
+	).AnyTimes()
+
+	err := s.nDCActivityStateReplicator.SyncActivitiesState(context.Background(), request)
```
Comment (xwduan) [inline service/history/ndc/activity_state_replicator_test.go:644]: Tests are copied from original test case and replaced with new api and request.

File: proto/internal/temporal/server/api/historyservice/v1/request_response.proto [hunk 2]
```diff
@@ -466,7 +466,27 @@ message SyncActivityRequest {
...
+    google.protobuf.Timestamp scheduled_time = 3 [(gogoproto.stdtime) = true];
+    int64 started_event_id = 4;
+    google.protobuf.Timestamp started_time = 5 [(gogoproto.stdtime) = true];
+    google.protobuf.Timestamp last_heartbeat_time = 6 [(gogoproto.stdtime) = true];
+    temporal.api.common.v1.Payloads details = 7;
+    int32 attempt = 8;
+    temporal.api.failure.v1.Failure last_failure = 9;
+    string last_worker_identity = 10;
+    temporal.server.api.history.v1.VersionHistory version_history = 11;
+    temporal.server.api.workflow.v1.BaseExecutionInfo base_execution_info = 12;
```
Comment (wxing1292) [inline proto/internal/temporal/server/api/historyservice/v1/request_response.proto:489]: you do not need this right?

Comment (xwduan) [reply proto/internal/temporal/server/api/historyservice/v1/request_response.proto:489]: I think we don't need this for now. I have not found any usage of this in current code. I copied this field from the original [request](https://github.com/temporalio/temporal/blob/d7fbf91cadb352da1042bc47b8cda4daaa0c8dcd/proto/internal/temporal/server/api/historyservice/v1/request_response.proto#L468).

File: service/history/ndc/activity_state_replicator.go [hunk 3]
```diff
@@ -195,23 +323,9 @@ func (r *ActivityStateReplicatorImpl) SyncActivityState(
 	if _, err := workflow.NewTimerSequence(
 		mutableState,
 	).CreateNextActivityTimer(); err != nil {
```
Comment (wxing1292) [inline service/history/ndc/activity_state_replicator.go:325]: this function should be invoked if 
* there is >= 1 activity updated
* just before updating mutable state?

File: service/history/ndc/activity_state_replicator.go [hunk 4]
```diff
@@ -126,64 +130,188 @@ func (r *ActivityStateReplicatorImpl) SyncActivityState(
...
+			},
+			mutableState,
+			syncActivityInfo,
+		)
+		if err != nil {
+			return err
+		}
+		if applied == true {
+			anyEventApplied = true
+		}
```
Comment (wxing1292) [inline service/history/ndc/activity_state_replicator.go:234]: anyEventApplied |= applied

File: service/history/ndc/activity_state_replicator.go [hunk 5]
```diff
@@ -126,64 +130,188 @@ func (r *ActivityStateReplicatorImpl) SyncActivityState(
...
+	)
+}
+
+func (r *ActivityStateReplicatorImpl) syncSingleActivityState(
+	workflowKey *definition.WorkflowKey,
+	mutableState workflow.MutableState,
+	activitySyncInfo *historyservice.ActivitySyncInfo,
+) (applied bool, retError error) {
+	scheduledEventID := activitySyncInfo.GetScheduledEventId()
+	shouldApply, err := r.testVersionHistory(
```
Comment (wxing1292) [inline service/history/ndc/activity_state_replicator.go:262]: what if 50% of acitvity can be applied and other 50% cannot?
current behavior will just return err?

Comment (wxing1292) [reply service/history/ndc/activity_state_replicator.go:262]: i believe 
1. input validation that all activties belong to the same version history branch
2. before apply any activity, making sure ^ version history branch & current history branch does not diverge
3. split activities into 2 categories
  a. category one: activities with event ID already exists on current version history branch
  b. category two: activities with event ID NOT exists on current version history branch

for 3.b definitely error out with resend error
for 3.a, **can** be applied before return 3.b error

Comment (xwduan) [reply service/history/ndc/activity_state_replicator.go:267]: > what if 50% of acitvity can be applied and other 50% cannot?
current behavior will just return err?
Yeah, current implementation will return error if any activity failed within the batch. We can improve it if necessary in the future.

File: proto/internal/temporal/server/api/historyservice/v1/request_response.proto [hunk 6]
```diff
@@ -466,7 +466,26 @@ message SyncActivityRequest {
...
     temporal.server.api.workflow.v1.BaseExecutionInfo base_execution_info = 15;
+}
 
+message SyncActivitiesRequest {
+    string namespace_id = 1;
+    string workflow_id = 2;
+    string run_id = 3;
+    repeated ActivitySyncInfo activities_info = 4;
+}
+message ActivitySyncInfo {
```
Comment (yux0) [inline proto/internal/temporal/server/api/historyservice/v1/request_response.proto:477]: nit: empty line

File: service/history/ndc/activity_state_replicator.go [hunk 7]
```diff
@@ -126,69 +130,109 @@ func (r *ActivityStateReplicatorImpl) SyncActivityState(
...
-				NamespaceID: request.GetNamespaceId(),
-				WorkflowID:  request.GetWorkflowId(),
-				RunID:       request.GetRunId(),
-			},
-			VisibilityTimestamp: eventTime,
-			EventID:             request.GetScheduledEventId(),
-			Version:             request.GetVersion(),
-			Attempt:             request.GetAttempt(),
-		})
+	return executionContext.UpdateWorkflowExecutionWithNew(
```
Comment (yux0) [inline service/history/ndc/activity_state_replicator.go:167]: not sure if we should check the workflow running state

File: service/history/shard/engine.go [hunk 8]
```diff
@@ -88,6 +88,7 @@ type (
 		ReplicateWorkflowState(ctx context.Context, request *historyservice.ReplicateWorkflowStateRequest) error
 		SyncShardStatus(ctx context.Context, request *historyservice.SyncShardStatusRequest) error
 		SyncActivity(ctx context.Context, request *historyservice.SyncActivityRequest) error
+		SyncActivities(ctx context.Context, request *historyservice.SyncActivitiesRequest) error
```
[the remark this answers is not in the document]
Comment (xwduan) [reply service/history/shard/engine.go:91]: Yeah, the plan is to deprecate the SyncActivity().

File: service/history/ndc/activity_state_replicator.go (hunk 7 above)
Comment (xwduan) [reply service/history/ndc/activity_state_replicator.go:167]: I checked the original implementation, seems we did not check running state. What you mean is that we should make sure workflow is in running state before we UpdateWorkflowExecution?

**Review by yux0 — APPROVED** [review]

**Pull request merged**